### ISSUE TITLE
Enhancement: [`no-shadow`] Disallow shadowing globally defined variables?

### ISSUE THREAD

**Issue by brandongit2** [typescript-eslint/typescript-eslint#10847]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [x] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [x] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/no-shadow

### Description

It seems wrong to be able to do this:

```ts
const String = 3;
const str = String('hello'); // <-- "This expression is not callable."
```

This _feels_ common enough that there should be a lint rule for it, but I've dug around and haven't found one yet. The closest one is ESLint's [original `no-shadow` rule](https://eslint.org/docs/latest/rules/no-shadow#builtinglobals) with `builtinGlobals: true`, but obviously that doesn't use typed linting.

It would be nice if typescript-eslint's `no-shadow` could detect shadowing of global variables.

### Fail

```TypeScript
const String = 3;
const str = String('hello'); // <-- "This expression is not callable."
```

### Pass

```TypeScript
const str = String('hello');
```

**Comment by bradzacher** [comment]
> but obviously that doesn't use typed linting.

The extension rule is not type-aware - so our version of the rule is no smarter than the base rule and it won't ever use type-aware linting because doing so prevents many people from using the extension rule.

The `builtinGlobals` option is powered by the globals config you declare in eslint. It's a pretty decent approximation of things for the most part - for example it will catch your example of `String`.

What cases does it not cover for you?

**Comment by brandongit2** [comment]
I suppose `builtinGlobals` is good enough. Thanks for the response!
